### PR TITLE
ICP-4838 Add a timed initial state check to Z-Wave locks

### PR DESCRIPTION
**Pull request by dkirker**
Verify that we have received the initial state of the lock 2 minutes after it has been installed and at 2 minute intervals until we have received the initial battery and lock states.

Sometimes the user can get in to a situation where a lock may not receive packets requesting the initial states, or we never send the request for the initial states.

### CODE REVIEW

File: devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy [hunk 1]
```diff
@@ -102,6 +102,30 @@ import physicalgraph.zwave.commands.usercodev1.*
...
+		// We might have called updated() or configure() at some point but not have received a reply, so don't flood the network
+		if (!state.lastLockDetailsQuery || secondsPast(state.lastLockDetailsQuery, 2 * 60)) {
+			def actions = updated()
+
+			if (actions) {
+				sendHubCommand(actions.toHubAction())
+			}
+		}
+
+		runIn(120, installedCheck)
```
Comment (tpmanley) [inline devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy]: Did you test with the DTH running locally? I believe you need to add a third argument `[forceForLocallyExecuting: true]` to get this to run if the DTH is running locally.

Comment (dkirker) [reply devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy]: @tpmanley I will give this a try. Thanks for the heads up! The account I was using with the Connect Home I have doesn't have the account permissions to set the runLocally flag.

Comment (dkirker) [reply devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy]: @tpmanley Update incoming with that change!

**Comment by gonella** [comment]
Hi @dkirker, do you know if this fix could help on this issue? 

https://github.com/SmartThingsCommunity/SmartThingsPublic/issues/3046

File: devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy [hunk 2]
```diff
@@ -102,6 +102,34 @@ import physicalgraph.zwave.commands.usercodev1.*
...
+ * If not, verify that we have at least requested them or request them,
+ * and check again.
+ */
+def scheduleInstalledCheck() {
+	runIn(120, installedCheck, [forceForLocallyExecuting: true])
+}
+
+def installedCheck() {
+	if (device.currentState("lock") && device.currentState("battery")) {
+		unschedule("installedCheck")
```
Comment (tpmanley) [inline devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:120]: Do you need to unschedule? It looks like the function is only called when the `runIn` fires so I don't think there will be any outstanding schedule that needs to be unscheduled.

**Review by tpmanley — APPROVED** [review]

File: devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy (hunk 2 above)
Comment (dkirker) [reply devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:120]: @tpmanley Probably not. I thought I had remembered hearing that was best practice, but I could be making it up. If anyone thinks it should be removed I can def do that. What I haven't tried is to see if it is possible to schedule multiple instances (I've seen a few duplicate log.debug from installedCheck), and if calling unscheduled once unschedules them all. Or if that is even possible.

**Review by kianooshkarami — APPROVED** [review]

**Comment by dkirker** [comment]
@workingmonk Alright, so this is def ready to go to staging, and I will coordinate with Steven C. to test the rest of this week. I am confident the impact is low.

**Pull request merged**